chain/chain/src/test_utils/kv_runtime.rs

```diff
@@ -417,6 +417,10 @@ impl EpochManagerAdapter for MockEpochManager {
         Ok(self.num_shards)
     }
 
+    fn shard_ids(&self, _epoch_id: &EpochId) -> Result<Vec<ShardId>, EpochError> {
```

Comment:
Allocating here runs a risk of making some hot loops that check for shard_ids quite a bit more expensive. I would advocate for a `SmallVec` or perhaps returning an `impl Iterator<Item=ShardId>` (insert the result wrapper where you think it is more appropriate.)

Reply:
> perhaps returning an impl Iterator<Item=ShardId> (insert the result wrapper where you think it is more appropriate.)

This is a trait method so AFAIU, it cannot return an `iterator`.  Is there some way to doing that in rust now?

> Allocating here runs a risk of making some hot loops that check for shard_ids quite a bit more expensive. I would advocate for a `SmallVec`

Are you worried about the performance for the test code or the production code as well?  I am asking because your comment is on the test code.  

We could choose to return a `SmallVec` or we could alternatively add an additional function `fn is_valid_shard_id()` to handle those use cases.  

I am also wondering if we might be doing some premature optimisation.  Do you have specific checks in mind that will call this in a hot loop?

Reply:
> Are you worried about the performance for the test code or the production code as well? I am asking because your comment is on the test code.

My concern is largely with the production code. e.g. chain/chain/src/chain.rs calls this function frequently, and I don’t have a good intuition how frequently some of the functions there get called (especially those that are more nebulous like the `has_all_receipts`). Anyway, I just grepped for the first definition and wrote my comment there.

---

> This is a trait method so AFAIU, it cannot return an iterator. Is there some way to doing that in rust now?

Yeah traits make `impl Iterator` harder. Still possible with associated types like so:

```
trait Foo {
    type ShardIdIterator = ...;
    fn shard_ids(&self) -> Self::ShardIdIterator { ... }
}
```

but unfortunately that requires an [unstable feature](https://github.com/rust-lang/rust/issues/63063) if you can’t name the type being returned easily.

---

> I am also wondering if we might be doing some premature optimisation. Do you have specific checks in mind that will call this in a hot loop?

A fair point. It is unfortunate we do not have a good automated performance evaluation tool, and it is also true we’ve landed other major changes to performance sensitive areas in the past without too much ceremony.